### ISSUE TITLE
Rayon uses a lot of CPU when there's not a lot of work to do

### ISSUE THREAD

**Issue by jrmuizel** [rayon-rs/rayon#642]
The following program uses about 30% of the CPU on 4 core (8 HT) Linux and Mac machines
```rust
fn main() {
    while true {
        std::thread::sleep(std::time::Duration::from_millis(10));
        rayon::spawn(move || {  } );
    }
}
```

Reducing the sleep duration to 1ms pushes the CPU usage up to 200%.

**Comment by cuviper** [comment]
See also #576 and #614.

**Comment by jrmuizel** [comment]
A profile on macOS shows most of the time being spent in sched_yield

**Comment by jrmuizel** [comment]
Also, a similar program written using the `dispatch` crate uses about 3-4% cpu at the 1ms duration vs the 200% that rayon uses.

**Comment by jrmuizel** [comment]
@lukewagner here's an example of some of the bad.

**Comment by jrmuizel** [comment]
Rewriting this using crossbeam-channel produces an executable that only uses 4% cpu. 

```rust
    let (s, r) = unbounded();

    for i in 0..8 {
        let r = r.clone();
        thread::spawn( move || loop {
            let m: i32 = r.recv().unwrap();
        });
    }

    while true {
        std::thread::sleep(std::time::Duration::from_millis(1));
        s.send(0);
    }
```

Perhaps rayon should use the same wakeup infrastructure that crossbeam-channel uses?

**Comment by ishitatsuyuki** [comment]
From https://github.com/rayon-rs/rayon/blob/master/rayon-core/src/sleep/README.md, it seems rayon will repetitively yield to the operating system before a thread goes to sleep. How does this work wrt to syscall overhead? I think yielding to the scheduler is almost as expensive as blocking on a condition variable?

**Comment by cuviper** [comment]
> I think yielding to the scheduler is almost as expensive as blocking on a condition variable?

The net performance of blocking has to include the cost of the syscall to wake it up too. Even if the raw cost were the same, there's also a latency difference in fully blocking.

The `life` demo added in #603 shows this trade-off pretty well. Currently I get:

```
$ cargo run --release -- life bench
    Finished release [optimized] target(s) in 0.05s
     Running `/home/jistone/rust/rayon/target/release/rayon-demo life bench`
  serial:  114948948 ns
parallel:   32975762 ns -> 3.49x speedup
par_bridge:  631383430 ns -> 0.18x speedup

$ cargo run --release -- life play
    Finished release [optimized] target(s) in 0.05s
     Running `/home/jistone/rust/rayon/target/release/rayon-demo life play`
  serial: 60.00 fps
    cpu usage: 16.2%
parallel: 60.00 fps
  cpu usage: 45.1%
par_bridge: 60.00 fps
  cpu usage: 366.8%
```

If I change `ROUNDS_UNTIL_SLEEPY` and `ROUNDS_UNTIL_ASLEEP` to 1:

```
$ cargo run --release -- life bench
    Finished release [optimized] target(s) in 0.05s
     Running `/home/jistone/rust/rayon/target/release/rayon-demo life bench`
  serial:  111999829 ns
parallel:   40216218 ns -> 2.78x speedup
par_bridge:  625470343 ns -> 0.18x speedup

$ cargo run --release -- life play
    Finished release [optimized] target(s) in 0.05s
     Running `/home/jistone/rust/rayon/target/release/rayon-demo life play`
  serial: 60.00 fps
    cpu usage: 15.9%
parallel: 60.00 fps
  cpu usage: 33.8%
par_bridge: 60.00 fps
  cpu usage: 350.2%
```

So not as much speedup, but lower cpu usage.

**Comment by ishitatsuyuki** [comment]
A few links to what might making the problem worse:
- Usage of notify_all() (#394) means that we pay the condition variable penalty more than what we expect. A hypothesis might be that keeping the threads awake is fast because in that way `notify_all()` costs less.
- Rayon creates threads eagerly (#576). Threads are best when they're saturating CPU (in other words, "just enough"), otherwise it's more context switch = worse.

**Comment by nikomatsakis** [comment]
I've been working on a rewrite of the sleep system that I think will address this problem. I opened https://github.com/rayon-rs/rayon/pull/691 with an immediate refactoring, but I'm working on a more extensive rewrite. I hope to open an RFC in a few days with a sketch of the algorithm plus a few alternatives I'd like to evaluate.

The gist of it is this:

- In the new system, we almost never wake up more one thread at a time (the exception is when terminating, which wakes all threads so they can quit).
- When producing new work, we can try to avoid waking threads at all if we judge there to be idle threads that are currently awake.
- When signalling latches, we only wake up the one thread that is actually blocked on the latch (if any).

We'll have to benchmark it, of course, and we'll probably want to tune things.

**Comment by nikomatsakis** [comment]
Opened: https://github.com/rayon-rs/rfcs/pull/5 with a fairly complete description of what I have in mind. I've not implemented all of that yet, but I'm getting there.

**Comment by nikomatsakis** [comment]
Using the algorithm from the RFC seems to reduce this example to taking negligible CPU time:

```
> time -p /home/nmatsakis/versioned/rayon/target/release/rayon-demo noop --iters 1000
real 10.15
user 0.06
sys 0.04
> time -p /home/nmatsakis/versioned/rayon/target/release/rayon-demo noop --iters 1000 --sleep 1
real 1.09
user 0.05
sys 0.03
```

**Comment by djg** [comment]
I've been experimenting with this case where work is spawned every millisecond:

```rust
fn main() {
    while true {
        std::thread::sleep(std::time::Duration::from_millis(1));
        rayon::spawn(move || {  } );
    }
}
```

With the current release of `rayon`, all my HTs are pegged:

<img width="640" alt="Current Rayon" src="https://user-images.githubusercontent.com/156077/64658295-534a7d00-d47a-11e9-9671-ec444146822a.png">

With the `latch-target-thread` branch, CPU usage is massively reduced:

<img width="640" alt="Rayon latch-target-thread branch" src="https://user-images.githubusercontent.com/156077/64658347-85f47580-d47a-11e9-9eaf-e8d8cc4e2ddb.png">

Excited to see where this goes.

**Comment by jrmuizel** [comment]
Here's a screenshot from a GPUview profile of WebRender using Rayon.

It shows 2ms of spinning 2 cpus just starting and stopping the threads.

![rayon](https://user-images.githubusercontent.com/332653/79152585-79325a00-7d9a-11ea-9314-e7b66ec38fa0.PNG)

**Comment by cuviper** [comment]
@jrmuizel would it be possible for you to gather a similar profile with #746?

**Comment by jrmuizel** [comment]
Yeah. I'm trying that now.

**Comment by jrmuizel** [comment]
Here's a similar situation with #746 
![rayon2](https://user-images.githubusercontent.com/332653/79167803-50b85900-7db6-11ea-962a-c5d90ff50992.PNG)

It looks like we're spinning for less time which is good.

I haven't read through #746 but wouldn't it be better for us to wait on something instead of having 2 cores jumping between 4 thread spinning on `yield()`?

**Comment by cuviper** [comment]
We do ultimately wait on condvars, but there's some spinning on `yield()` first, in hopes of reducing latency. You can see some of that effect in [my earlier comment](https://github.com/rayon-rs/rayon/issues/642#issuecomment-521812112), where I cut down the number of spinning rounds. Still, there may be room to fine-tune this further in #746.

**Comment by cuviper** [comment]
I'm going to close this issue now that 1.4.0 is published.

@hniksic could you open a new issue for your case? There are definitely some scalability issues with `par_bridge()` that we should discuss separately.

**Comment by hniksic** [comment]
@cuviper I was loath to open an issue since I couldn't provide a minimal example that reproduces the bug, and I tried. If you think the above description would be useful enough as an issue of its own, I'll gladly make one.

**Comment by cuviper** [comment]
@hniksic Issues are not a scarce resource. :slightly_smiling_face: We can at least discuss the problem in the abstract, and there might be other people that want to chime in, who may have something more reproducible.

**Comment by hniksic** [comment]
@cuviper Thanks! I've now submitted #795.
